# Phase 1: Data Cleaning, Pre-Processing

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = pd.read_excel('Sales_Contribution.xlsx', names = ['Sr', 'Comp', 'Code', 'Item Description','', 'Packing', 'Stock', 'Qnty', 'Sales', 'Contri%'], skiprows=4)
ndf = df
ndf

Sr  Comp         Code  \
0     ----------------------------------------------...   NaN          NaN   
1                                                     1  KWA         60496   
2                                                     2  GENR        21986   
3                                                     3  DIN         57932   
4                                                     4  FDC         15063   
...                                                 ...   ...          ...   
3397                                               3397  DAYL        65416   
3398                                               3398  IPCA        42948   
3399  ----------------------------------------------...   NaN          NaN   
3400                                                NaN   NaN          NaN   
3401                                                NaN   NaN  TOTAL SALES   

         Item Description              Packing  Stock  \
0                     NaN         NaN      NaN    NaN   
1        ICE CREAM (RS10)         NaN      25G  472.0   
2                  HALLS          NaN      1PC  235.0   
3       ICE CREAM (RS.20)         NaN     PACK  470.0   
4         ELECTRAL POWDER         NaN  21.80GM   14.0   
...                   ...         ...      ...    ...   
3397           ZREPAG 2MF         NaN      10T    5.0   
3398  REVELOL AM 25/2.5MG         NaN      10T    8.0   
3399                  NaN         NaN      NaN    NaN   
3400                  NaN         NaN      NaN    NaN   
3401                  NaN  1509983.04      NaN    NaN   

                               Qnty       Sales  Contri%  
0                               NaN         NaN      NaN  
1                               360     3600.00     0.24  
2                               182      182.00     0.01  
3                               158     3160.00     0.21  
4                               137     3335.95     0.22  
...                             ...         ...      ...  
3397                             -5     -800.00    -0.05  
3398                             -7     -997.50    -0.07  
3399                            NaN         NaN      NaN  
3400   Sales total of TOP  products  1509983.04   100.00  
3401                            NaN         NaN      NaN  

[3402 rows x 10 columns]

In [4]:
ndf.describe()

Stock         Sales      Contri%
count        1.00  3063.000000  3.371000e+03  2944.000000
mean   1509983.04     7.496246  8.958665e+02     0.067942
std           NaN    55.971754  2.600701e+04     1.842901
min    1509983.04    -1.000000 -9.975000e+02    -0.070000
25%    1509983.04     1.000000  1.180000e+02     0.010000
50%    1509983.04     2.000000  2.378700e+02     0.020000
75%    1509983.04     3.000000  5.233000e+02     0.040000
max    1509983.04  2374.000000  1.509983e+06   100.000000

In [5]:
ndf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3402 entries, 0 to 3401
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Sr                3400 non-null   object 
 1   Comp              3398 non-null   object 
 2   Code              3399 non-null   object 
 3   Item Description  3398 non-null   object 
 4                     1 non-null      float64
 5   Packing           3392 non-null   object 
 6   Stock             3063 non-null   float64
 7   Qnty              3356 non-null   object 
 8   Sales             3371 non-null   float64
 9   Contri%           2944 non-null   float64
dtypes: float64(4), object(6)
memory usage: 265.9+ KB


In [6]:
ndf.sample(10)

Sr  Comp   Code       Item Description     Packing  Stock Qnty  \
3078  3078  DR.M  58632      CALCIQUICK-D3 60K NaN    4CAP    3.0    1   
2309  2309  GENR   3236       ARROWROOT POWDER NaN    50GM    1.0    1   
1563  1563  HIMA  45351               SEPTILIN NaN     60T    1.0    2   
1733  1733  NUTR  59736  NUTRA FIZZ C 1000 TAB NaN   20TAB    3.0    2   
1156  1156  APEL  37643            P 500 SYRUP NaN    60ml    2.0    3   
689    689  IPCA  63894      CTD MT 6.25/50/40 NaN     10T    3.0    4   
37      37  PF     4999          BECOSULES CAP NaN     20C   17.0   34   
1110  1110  CIPL   4199             AZEE 250MG NaN     10T    6.0    3   
196    196  JENB  33259          NERVIJEN PLUS NaN     10C    1.0   10   
2638  2638  MEDL  37276    OSTIUM-K2 FORTE TAB NaN     15T    NaN    1   

        Sales  Contri%  
3078   137.84     0.01  
2309   110.00     0.01  
1563   405.00     0.03  
1733   738.00     0.05  
1156   143.13     0.01  
689    509.00     0.03  
37    1870.00     0.12  
1110   391.08     0.03  
196   1995.00     0.13  
2638   435.00     0.03

In [7]:
ndf.columns

Index(['Sr', 'Comp', 'Code', 'Item Description', '', 'Packing', 'Stock',
       'Qnty', 'Sales', 'Contri%'],
      dtype='object')

In [8]:
ndf.drop(['', 'Sr'], axis=1, inplace=True)
ndf

Comp         Code     Item Description  Packing  Stock  \
0      NaN          NaN                  NaN      NaN    NaN   
1     KWA         60496     ICE CREAM (RS10)      25G  472.0   
2     GENR        21986               HALLS       1PC  235.0   
3     DIN         57932    ICE CREAM (RS.20)     PACK  470.0   
4     FDC         15063      ELECTRAL POWDER  21.80GM   14.0   
...    ...          ...                  ...      ...    ...   
3397  DAYL        65416           ZREPAG 2MF      10T    5.0   
3398  IPCA        42948  REVELOL AM 25/2.5MG      10T    8.0   
3399   NaN          NaN                  NaN      NaN    NaN   
3400   NaN          NaN                  NaN      NaN    NaN   
3401   NaN  TOTAL SALES                  NaN      NaN    NaN   

                               Qnty       Sales  Contri%  
0                               NaN         NaN      NaN  
1                               360     3600.00     0.24  
2                               182      182.00     0.01  
3                               158     3160.00     0.21  
4                               137     3335.95     0.22  
...                             ...         ...      ...  
3397                             -5     -800.00    -0.05  
3398                             -7     -997.50    -0.07  
3399                            NaN         NaN      NaN  
3400   Sales total of TOP  products  1509983.04   100.00  
3401                            NaN         NaN      NaN  

[3402 rows x 8 columns]

In [9]:
ndf.dropna(subset=['Item Description'], inplace=True)
ndf

Comp   Code     Item Description  Packing  Stock Qnty    Sales  Contri%
1     KWA   60496     ICE CREAM (RS10)      25G  472.0  360  3600.00     0.24
2     GENR  21986               HALLS       1PC  235.0  182   182.00     0.01
3     DIN   57932    ICE CREAM (RS.20)     PACK  470.0  158  3160.00     0.21
4     FDC   15063      ELECTRAL POWDER  21.80GM   14.0  137  3335.95     0.22
5     P&G   53290    VICKS COUGH DROPS      1PC  707.0  127   127.00     0.01
...    ...    ...                  ...      ...    ...  ...      ...      ...
3394  ZHL   62391      STEOMIN OS TAB    10 TAB    1.0   -3  -570.00    -0.04
3395  CIPL   1708             AMANTREL      15C    4.0   -3  -733.47    -0.05
3396  DAYL  59958       ROZUTION T TAB    10TAB    4.0   -4  -516.00    -0.03
3397  DAYL  65416           ZREPAG 2MF      10T    5.0   -5  -800.00    -0.05
3398  IPCA  42948  REVELOL AM 25/2.5MG      10T    8.0   -7  -997.50    -0.07

[3398 rows x 8 columns]

*The Contri% column seems inconsistent with the available data, hence dropping it entirely seems a good idea*

In [11]:
ndf.drop('Contri%', axis = 1, inplace=True)
ndf

Comp   Code     Item Description  Packing  Stock Qnty    Sales
1     KWA   60496     ICE CREAM (RS10)      25G  472.0  360  3600.00
2     GENR  21986               HALLS       1PC  235.0  182   182.00
3     DIN   57932    ICE CREAM (RS.20)     PACK  470.0  158  3160.00
4     FDC   15063      ELECTRAL POWDER  21.80GM   14.0  137  3335.95
5     P&G   53290    VICKS COUGH DROPS      1PC  707.0  127   127.00
...    ...    ...                  ...      ...    ...  ...      ...
3394  ZHL   62391      STEOMIN OS TAB    10 TAB    1.0   -3  -570.00
3395  CIPL   1708             AMANTREL      15C    4.0   -3  -733.47
3396  DAYL  59958       ROZUTION T TAB    10TAB    4.0   -4  -516.00
3397  DAYL  65416           ZREPAG 2MF      10T    5.0   -5  -800.00
3398  IPCA  42948  REVELOL AM 25/2.5MG      10T    8.0   -7  -997.50

[3398 rows x 7 columns]

*Now, I will find the items with 0 potential sales, or negligible Qnty, and then decide what to do.*

In [15]:
ndf[ndf.Sales.isna()]

Comp   Code           Item Description Packing  Stock Qnty  Sales
3326  ALKM  64110            NERVMAX D 50/20     10C    2.0  NaN    NaN
3327  ALCH  39279         PHYTORELIEF-CC TAB     10T    NaN  NaN    NaN
3330  1225  62685                INSTAAHEAL      10T    1.0  NaN    NaN
3334  CADP    349             ACILOC RD 20MG   10TAB    2.0  NaN    NaN
3336  DR.R   3572               ATARAX DROPS    15ML    3.0  NaN    NaN
3337  DAYL  60174       DAYCLAZIDE MP 40 TAB  10 TAB    1.0  NaN    NaN
3338  ELDG  65402                DISORB M 50     10T    2.0  NaN    NaN
3339  CIPL  14683                 DYTOR 40MG     10T    1.0  NaN    NaN
3340  EPL   65494               VERIHEAL OIL   100ML    NaN  NaN    NaN
3341  ER    65275                GLUXIT M 10   10TAB    1.0  NaN    NaN
3342  GENR  65452                AMNIO SHETS       1    NaN  NaN    NaN
3343  JIVI  63474                 C SPIN TAB     20T    1.0  NaN    NaN
3344  IPCA   2856     AQUASURGE MAX EYE DROP    10ml    1.0  NaN    NaN
3345  INTA  22622                 HIFENAC P      15T    2.0  NaN    NaN
3346  MACL  55747        ZEDOCEF DXL 200 TAB     10T    2.0  NaN    NaN
3347  IMP    3985              AXE BRAND OIL    10ML    1.0  NaN    NaN
3348  HESH   6124     BORIC ACID POWDER-HESH    50GM    4.0  NaN    NaN
3350  HE    58596            LULICONAZ CREAM    10GM    2.0  NaN    NaN
3353  GERM  55370       XYLOMIST 0.1MG DROPS    10ML    2.0  NaN    NaN
3357  GENR  61354          LIBERTY PANT (XL)     1PC    1.0  NaN    NaN
3358  PF    65128              ETORIZER 60MG     15T    2.0  NaN    NaN
3359  PGON  39341  PIGEON SI. NIPPLE SHIELD      1PC    4.0  NaN    NaN
3360  NOVA  65513              GALVUS OD 100     15T    NaN  NaN    NaN
3363  NVCD  63627      NIVEA BODY LOTION GEL    75ML    3.0  NaN    NaN
3364  ZKL   58614               DAILYGLIM-M2   10TAB   11.0  NaN    NaN
3365  WALN  65511               MELTOFAST 10     10T    1.0  NaN    NaN
3366  UNP   65475                    MYRACID     10C    1.0  NaN    NaN
3368  SYST  49442          TERBEST 250MG TAB      7T    2.0  NaN    NaN

*Filling these with 0 seems a better option, but first, I will find whether there are other missing values*

In [18]:
ndf[ndf['Item Description'].isna()]

Empty DataFrame
Columns: [Comp, Code, Item Description, Packing, Stock, Qnty, Sales]
Index: []

In [19]:
ndf.isna().sum()

Comp                  0
Code                  0
Item Description      0
Packing               6
Stock               335
Qnty                 43
Sales                28
dtype: int64

In [20]:
ndf[ndf['Qnty'].isna()]

Comp   Code           Item Description Packing  Stock Qnty   Sales
3326  ALKM  64110            NERVMAX D 50/20     10C    2.0  NaN     NaN
3327  ALCH  39279         PHYTORELIEF-CC TAB     10T    NaN  NaN     NaN
3328  ALGD   1417                ALMOX 500MG     15c   12.0  NaN  181.06
3329  ALGD  36309                 OMEE-D CAP     20C    3.0  NaN   70.00
3330  1225  62685                INSTAAHEAL      10T    1.0  NaN     NaN
3331  CIGD  46756                  SPASMONIL     10T   42.0  NaN    5.80
3332  CIGD  47705                 SUHAGRA-50      4T    1.0  NaN  133.08
3333  CIGD  10147                  COF-Q TAB     10T   21.0  NaN    9.24
3334  CADP    349             ACILOC RD 20MG   10TAB    2.0  NaN     NaN
3335  DR.   65335       DR.RASHEL NOSE STRIP      1P    NaN  NaN   56.00
3336  DR.R   3572               ATARAX DROPS    15ML    3.0  NaN     NaN
3337  DAYL  60174       DAYCLAZIDE MP 40 TAB  10 TAB    1.0  NaN     NaN
3338  ELDG  65402                DISORB M 50     10T    2.0  NaN     NaN
3339  CIPL  14683                 DYTOR 40MG     10T    1.0  NaN     NaN
3340  EPL   65494               VERIHEAL OIL   100ML    NaN  NaN     NaN
3341  ER    65275                GLUXIT M 10   10TAB    1.0  NaN     NaN
3342  GENR  65452                AMNIO SHETS       1    NaN  NaN     NaN
3343  JIVI  63474                 C SPIN TAB     20T    1.0  NaN     NaN
3344  IPCA   2856     AQUASURGE MAX EYE DROP    10ml    1.0  NaN     NaN
3345  INTA  22622                 HIFENAC P      15T    2.0  NaN     NaN
3346  MACL  55747        ZEDOCEF DXL 200 TAB     10T    2.0  NaN     NaN
3347  IMP    3985              AXE BRAND OIL    10ML    1.0  NaN     NaN
3348  HESH   6124     BORIC ACID POWDER-HESH    50GM    4.0  NaN     NaN
3349  HE    65298          IBUBROOK PLUS TAB     15T    4.0  NaN    4.72
3350  HE    58596            LULICONAZ CREAM    10GM    2.0  NaN     NaN
3351  HE    60310            HEALTUSS COFGEL     10C   24.0  NaN   44.00
3352  GGOC  58668              PRIMONT-N TAB   10TAB    6.0  NaN   21.52
3353  GERM  55370       XYLOMIST 0.1MG DROPS    10ML    2.0  NaN     NaN
3354  GENR  23014         HIPHOP NOSE STRIPS     3PC    3.0  NaN   83.32
3355  GENR  24657                  JAI KAJAL   2.5GM   14.0  NaN   50.00
3356  GENR  62672              NISHA MEHANDI     10G    3.0  NaN   45.00
3357  GENR  61354          LIBERTY PANT (XL)     1PC    1.0  NaN     NaN
3358  PF    65128              ETORIZER 60MG     15T    2.0  NaN     NaN
3359  PGON  39341  PIGEON SI. NIPPLE SHIELD      1PC    4.0  NaN     NaN
3360  NOVA  65513              GALVUS OD 100     15T    NaN  NaN     NaN
3361  OMEG   2205              AMPICLOX PLUS     10C    6.0  NaN   30.00
3362  OMEG  57521            MEFNAC-SPAS TAB     10T   13.0  NaN    8.00
3363  NVCD  63627      NIVEA BODY LOTION GEL    75ML    3.0  NaN     NaN
3364  ZKL   58614               DAILYGLIM-M2   10TAB   11.0  NaN     NaN
3365  WALN  65511               MELTOFAST 10     10T    1.0  NaN     NaN
3366  UNP   65475                    MYRACID     10C    1.0  NaN     NaN
3367  TAN    3937                     AVIMAX     10T   16.0  NaN   10.00
3368  SYST  49442          TERBEST 250MG TAB      7T    2.0  NaN     NaN

*I got to know that stock column is missing 335 values, and filling all with 0 is a dumb decision, a better decision is to fill them with same value as Qnty.*

In [21]:
ndf['Stock'].fillna(ndf['Qnty'], inplace=True)
ndf

C:\Users\Jenil\AppData\Local\Temp\ipykernel_11704\235435075.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  ndf['Stock'].fillna(ndf['Qnty'], inplace=True)
C:\Users\Jenil\AppData\Local\Temp\ipykernel_11704\235435075.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[60 33 12 12 12 11 10 10 10 9 7 6 6 6 6 6 6 6 6 6 6 6 6 5 5 5 5 5 5 5 5 5
 4 4 4 4 4 4 4 4 4 4 4 4 4 4 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
 3 3 3 3 2 2 2 2 2 2 2 2 2 2 

Comp   Code     Item Description  Packing  Stock Qnty    Sales
1     KWA   60496     ICE CREAM (RS10)      25G  472.0  360  3600.00
2     GENR  21986               HALLS       1PC  235.0  182   182.00
3     DIN   57932    ICE CREAM (RS.20)     PACK  470.0  158  3160.00
4     FDC   15063      ELECTRAL POWDER  21.80GM   14.0  137  3335.95
5     P&G   53290    VICKS COUGH DROPS      1PC  707.0  127   127.00
...    ...    ...                  ...      ...    ...  ...      ...
3394  ZHL   62391      STEOMIN OS TAB    10 TAB    1.0   -3  -570.00
3395  CIPL   1708             AMANTREL      15C    4.0   -3  -733.47
3396  DAYL  59958       ROZUTION T TAB    10TAB    4.0   -4  -516.00
3397  DAYL  65416           ZREPAG 2MF      10T    5.0   -5  -800.00
3398  IPCA  42948  REVELOL AM 25/2.5MG      10T    8.0   -7  -997.50

[3398 rows x 7 columns]

In [22]:
ndf.isna().sum()

Comp                 0
Code                 0
Item Description     0
Packing              6
Stock                5
Qnty                43
Sales               28
dtype: int64

In [23]:
ndf[ndf.Qnty.isna()]

Comp   Code           Item Description Packing  Stock Qnty   Sales
3326  ALKM  64110            NERVMAX D 50/20     10C    2.0  NaN     NaN
3327  ALCH  39279         PHYTORELIEF-CC TAB     10T    NaN  NaN     NaN
3328  ALGD   1417                ALMOX 500MG     15c   12.0  NaN  181.06
3329  ALGD  36309                 OMEE-D CAP     20C    3.0  NaN   70.00
3330  1225  62685                INSTAAHEAL      10T    1.0  NaN     NaN
3331  CIGD  46756                  SPASMONIL     10T   42.0  NaN    5.80
3332  CIGD  47705                 SUHAGRA-50      4T    1.0  NaN  133.08
3333  CIGD  10147                  COF-Q TAB     10T   21.0  NaN    9.24
3334  CADP    349             ACILOC RD 20MG   10TAB    2.0  NaN     NaN
3335  DR.   65335       DR.RASHEL NOSE STRIP      1P    NaN  NaN   56.00
3336  DR.R   3572               ATARAX DROPS    15ML    3.0  NaN     NaN
3337  DAYL  60174       DAYCLAZIDE MP 40 TAB  10 TAB    1.0  NaN     NaN
3338  ELDG  65402                DISORB M 50     10T    2.0  NaN     NaN
3339  CIPL  14683                 DYTOR 40MG     10T    1.0  NaN     NaN
3340  EPL   65494               VERIHEAL OIL   100ML    NaN  NaN     NaN
3341  ER    65275                GLUXIT M 10   10TAB    1.0  NaN     NaN
3342  GENR  65452                AMNIO SHETS       1    NaN  NaN     NaN
3343  JIVI  63474                 C SPIN TAB     20T    1.0  NaN     NaN
3344  IPCA   2856     AQUASURGE MAX EYE DROP    10ml    1.0  NaN     NaN
3345  INTA  22622                 HIFENAC P      15T    2.0  NaN     NaN
3346  MACL  55747        ZEDOCEF DXL 200 TAB     10T    2.0  NaN     NaN
3347  IMP    3985              AXE BRAND OIL    10ML    1.0  NaN     NaN
3348  HESH   6124     BORIC ACID POWDER-HESH    50GM    4.0  NaN     NaN
3349  HE    65298          IBUBROOK PLUS TAB     15T    4.0  NaN    4.72
3350  HE    58596            LULICONAZ CREAM    10GM    2.0  NaN     NaN
3351  HE    60310            HEALTUSS COFGEL     10C   24.0  NaN   44.00
3352  GGOC  58668              PRIMONT-N TAB   10TAB    6.0  NaN   21.52
3353  GERM  55370       XYLOMIST 0.1MG DROPS    10ML    2.0  NaN     NaN
3354  GENR  23014         HIPHOP NOSE STRIPS     3PC    3.0  NaN   83.32
3355  GENR  24657                  JAI KAJAL   2.5GM   14.0  NaN   50.00
3356  GENR  62672              NISHA MEHANDI     10G    3.0  NaN   45.00
3357  GENR  61354          LIBERTY PANT (XL)     1PC    1.0  NaN     NaN
3358  PF    65128              ETORIZER 60MG     15T    2.0  NaN     NaN
3359  PGON  39341  PIGEON SI. NIPPLE SHIELD      1PC    4.0  NaN     NaN
3360  NOVA  65513              GALVUS OD 100     15T    NaN  NaN     NaN
3361  OMEG   2205              AMPICLOX PLUS     10C    6.0  NaN   30.00
3362  OMEG  57521            MEFNAC-SPAS TAB     10T   13.0  NaN    8.00
3363  NVCD  63627      NIVEA BODY LOTION GEL    75ML    3.0  NaN     NaN
3364  ZKL   58614               DAILYGLIM-M2   10TAB   11.0  NaN     NaN
3365  WALN  65511               MELTOFAST 10     10T    1.0  NaN     NaN
3366  UNP   65475                    MYRACID     10C    1.0  NaN     NaN
3367  TAN    3937                     AVIMAX     10T   16.0  NaN   10.00
3368  SYST  49442          TERBEST 250MG TAB      7T    2.0  NaN     NaN

In [24]:
ndf[ndf.Stock.isna()]

Comp   Code      Item Description Packing  Stock Qnty  Sales
3327  ALCH  39279    PHYTORELIEF-CC TAB     10T    NaN  NaN    NaN
3335  DR.   65335  DR.RASHEL NOSE STRIP      1P    NaN  NaN   56.0
3340  EPL   65494          VERIHEAL OIL   100ML    NaN  NaN    NaN
3342  GENR  65452           AMNIO SHETS       1    NaN  NaN    NaN
3360  NOVA  65513         GALVUS OD 100     15T    NaN  NaN    NaN

*The null values in Qnty is a lot, and I will use regression to later deal with them, for now, let's remove other null values*

In [33]:
ndf.dropna(subset=['Stock', 'Sales'], inplace = True, how = 'all')

In [35]:
ndf[ndf.Stock.isna()]

Comp   Code      Item Description Packing  Stock Qnty  Sales
3335  DR.   65335  DR.RASHEL NOSE STRIP      1P    NaN  NaN   56.0

In [39]:
ndf.Stock.fillna(1,inplace=True)

C:\Users\Jenil\AppData\Local\Temp\ipykernel_11704\3707706496.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  ndf.Stock.fillna(1,inplace=True)


In [40]:
ndf[ndf.Stock.isna()]

Empty DataFrame
Columns: [Comp, Code, Item Description, Packing, Stock, Qnty, Sales]
Index: []

In [43]:
ndf.fillna({'Packing': '1 Pack'}, inplace = True)

In [45]:
ndf.isna().sum()

Comp                 0
Code                 0
Item Description     0
Packing              0
Stock                0
Qnty                39
Sales               24
dtype: int64

In [47]:
ndf.Sales.fillna(0, inplace = True)

C:\Users\Jenil\AppData\Local\Temp\ipykernel_11704\1280582001.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  ndf.Sales.fillna(0, inplace = True)


In [50]:
ndf.isna().sum()

Comp                 0
Code                 0
Item Description     0
Packing              0
Stock                0
Qnty                39
Sales                0
dtype: int64

In [52]:
ndf.to_excel('clean_salescontri.xlsx', index = False)

**Now, we can use ML to predict missing Qnty values**